File: string/454-longestPalindrome.cpp
#include<string>
#include<iostream>
#include<map>

using STR=std::string;
using MAP=std::map<char,int>;
MAP myCounter(const STR& s)
{
    MAP m{};
    for(auto c:s)
    {
        m[c]+=1;
    }
    return m;
}
int longestPalindrome(const STR& s)
{
    if(s.size()==1)
        return 1;
    MAP m=myCounter(s);
    int ans=0,odd=0;
    for(auto &[k,v]:m)
    {
        if((v&1)==0)  //& operator is lower than "==",so need the ()
            ans+=v;
        else
        {
            ans+=(v-1);
            odd=1;
        }
    }
    return ans+odd;
}

int main()
{
    STR s{"abccccdd"};
    std::cout<<longestPalindrome(s);
}

File: string/454-longestPalindrome.py
'''
Given a string s which consists of lowercase or uppercase letters, return the length of the longest palindrome that can be built with those letters.

Letters are case sensitive, for example, "Aa" is not considered a palindrome here.

 

Example 1:

Input: s = "abccccdd"
Output: 7
Explanation: One longest palindrome that can be built is "dccaccd", whose length is 7.
Example 2:

Input: s = "a"
Output: 1
Explanation: The longest palindrome that can be built is "a", whose length is 1.
'''
class Solution(object):
    def longestPalindrome(self, s):
        """
        :type s: str
        :rtype: int
        """
        if len(s)==1:
            return 1

        c=Counter(s)
        ans=0
        maxOdd=0
        odd=0
        for k,v in c.items():
            if not v&1:
                ans+=v
            else:
                ans+=v-1
                odd=1

        return ans+odd